diff -urN coreutils-5.2.1.orig/lib/userspec.c coreutils-5.2.1/lib/userspec.c
--- coreutils-5.2.1.orig/lib/userspec.c	2004-03-02 23:57:20.000000000 -0800
+++ coreutils-5.2.1/lib/userspec.c	2004-05-16 10:57:55.000000000 -0700
@@ -169,7 +169,7 @@
   separator = strchr (spec, ':');
 
   /* If there is no colon, then see if there's a `.'.  */
-  if (separator == NULL)
+  if (separator == NULL && (!getenv ("POSIXLY_CORRECT") || posix2_version() < 200112))
     {
       dot = strchr (spec, '.');
       /* If there's no colon but there is a `.', then first look up the
@@ -177,6 +177,11 @@
 	 If that fails, then we'll try again, but interpreting the `.'
 	 as a separator.  This is a compatible extension to POSIX, since
 	 the POSIX-required behavior is always tried first.  */
+	 if (dot)
+	   {
+	     if (! (posix2_version() < 200112))
+	       error (0, 0, _("`%s' syntax is obsolete; use `:' since this will be removed in the future"), spec_arg);
+	   }
     }
 
  retry:
