--- src/head.c.old	2004-01-27 23:19:56.697972992 -0500
+++ src/head.c	2004-01-27 23:34:01.301573728 -0500
@@ -995,14 +995,15 @@
 	    }
 	}
 
-      if (200112 <= posix2_version () && getenv ("POSIXLY_CORRECT"))
+      if (200112 <= posix2_version ())
 	{
-	  error (0, 0, _("`-%s' option is obsolete; use `-%c %.*s%.*s%s'"),
+	  error (0, 0, _("`-%s' option is obsolete; use `-%c %.*s%.*s%s' since this will be removed in the future"),
 		 n_string, count_lines ? 'n' : 'c',
 		 (int) (end_n_string - n_string), n_string,
 		 multiplier_char != 0, &multiplier_char,
 		 header_mode_option[header_mode]);
-	  usage (EXIT_FAILURE);
+	  if (getenv ("POSIXLY_CORRECT"))
+		  usage (EXIT_FAILURE);
 	}
 
       /* Append the multiplier character (if any) onto the end of
