--- src/nice.c.old	2004-01-27 23:24:40.927763496 -0500
+++ src/nice.c	2004-01-27 23:54:51.756475688 -0500
@@ -108,6 +108,8 @@
       if (s[0] == '-' && s[1] == '-' && ISDIGIT (s[2])
 	  && (posix2_version () < 200112 || !getenv ("POSIXLY_CORRECT")))
 	{
+	  if (!getenv ("POSIXLY_CORRECT"))
+		error (0, 0, _("`%s' option is obsolete; use `-n %s' since this will be removed in the future"), s, s+1);
 	  if (xstrtol (&s[2], NULL, 10, &adjustment, "") != LONGINT_OK)
 	    error (EXIT_FAILURE, 0, _("invalid option `%s'"), s);
 
@@ -119,6 +121,8 @@
 	       && (ISDIGIT (s[1]) || (s[1] == '+' && ISDIGIT (s[2])))
 	       && (posix2_version () < 200112 || !getenv ("POSIXLY_CORRECT")))
 	{
+	  if (!getenv ("POSIXLY_CORRECT"))
+		error (0, 0, _("`%s' option is obsolete; use `-n %s' since this will be removed in the future"), s, s);
 	  if (s[1] == '+')
 	    ++s;
 	  if (xstrtol (&s[1], NULL, 10, &adjustment, "") != LONGINT_OK)
