diff -ruN software-suspend-linux-2.6.7-rev12/kernel/power/Kconfig software-suspend-linux-2.6.7-rev13/kernel/power/Kconfig
--- software-suspend-linux-2.6.7-rev12/kernel/power/Kconfig	2004-07-13 16:57:31.000000000 +1000
+++ software-suspend-linux-2.6.7-rev13/kernel/power/Kconfig	2004-07-13 16:57:35.000000000 +1000
@@ -124,6 +124,15 @@
 		
 			  For normal usage, this option can be turned off.
 
+		config SOFTWARE_SUSPEND_ACPI
+			bool '   Use ACPI S4 support'
+			---help---
+			  This option enables calls to the ACPI code that handles entering
+			  the S4 (suspend-to-disk) state. Since some BIOSes have broken S4
+			  implementations, you might prefer to turn this off. On the other
+			  hand, ACPI wake events may fail to turn the computer back on if
+			  this option is disabled.
+
 		config SOFTWARE_SUSPEND_VARIATION_ANALYSIS
 			bool '   Variation Analysis'
 			---help---
diff -ruN software-suspend-linux-2.6.7-rev12/kernel/power/Makefile software-suspend-linux-2.6.7-rev13/kernel/power/Makefile
--- software-suspend-linux-2.6.7-rev12/kernel/power/Makefile	2004-07-13 16:57:31.000000000 +1000
+++ software-suspend-linux-2.6.7-rev13/kernel/power/Makefile	2004-07-13 16:57:35.000000000 +1000
@@ -5,7 +5,7 @@
 obj-$(CONFIG_SOFTWARE_SUSPEND_NULLWRITER)	+= nullwriter.o
 obj-$(CONFIG_SOFTWARE_SUSPEND_NFSWRITER)	+= nfswriter.o
 obj-$(CONFIG_SOFTWARE_SUSPEND_SWAPWRITER)	+= swapwriter.o block_io.o
-obj-$(CONFIG_SOFTWARE_SUSPEND2) += io.o memory_pool.o pagedir.o prepare_image.o proc.o range.o suspend2.o
+obj-$(CONFIG_SOFTWARE_SUSPEND2) += io.o memory_pool.o pagedir.o prepare_image.o proc.o range.o suspend2.o plugins.o
 obj-$(CONFIG_SOFTWARE_SUSPEND)	+= swsusp.o
 obj-$(CONFIG_PM_DISK)		+= disk.o pmdisk.o
 
diff -ruN software-suspend-linux-2.6.7-rev12/mm/highmem.c software-suspend-linux-2.6.7-rev13/mm/highmem.c
--- software-suspend-linux-2.6.7-rev12/mm/highmem.c	2004-07-13 16:57:31.000000000 +1000
+++ software-suspend-linux-2.6.7-rev13/mm/highmem.c	2004-07-13 16:57:35.000000000 +1000
@@ -26,6 +26,7 @@
 #include <linux/init.h>
 #include <linux/hash.h>
 #include <linux/highmem.h>
+#include <linux/suspend.h>
 #include <asm/pgalloc.h>
 #include <asm/tlbflush.h>
 
