diff -ruN software-suspend-core-2.0.0.97/include/linux/suspend-debug.h software-suspend-core-2.0.0.98/include/linux/suspend-debug.h
--- software-suspend-core-2.0.0.97/include/linux/suspend-debug.h	2004-07-09 23:13:15.000000000 +1000
+++ software-suspend-core-2.0.0.98/include/linux/suspend-debug.h	2004-07-09 23:13:21.000000000 +1000
@@ -5,7 +5,7 @@
 /* Solely for comparison with the version specific patch revision */
 #define SWSUSP_CORE_REVISION 0x201
 
-#define SWSUSP_CORE_VERSION "2.0.0.97"
+#define SWSUSP_CORE_VERSION "2.0.0.98"
 #define name_suspend "Software Suspend " SWSUSP_CORE_VERSION ": "
 
 /* Same length to ensure one overwrites the other */
diff -ruN software-suspend-core-2.0.0.97/kernel/power/process.c software-suspend-core-2.0.0.98/kernel/power/process.c
--- software-suspend-core-2.0.0.97/kernel/power/process.c	2004-07-09 23:13:15.000000000 +1000
+++ software-suspend-core-2.0.0.98/kernel/power/process.c	2004-07-09 23:13:22.000000000 +1000
@@ -583,7 +583,7 @@
 	return notifier_chain_unregister(&resume_notifier_list, nb);
 }
 
-inline int notify_resume(void)
+int notify_resume(void)
 {
 	return notifier_call_chain(&resume_notifier_list, 0, NULL);
 }
diff -ruN software-suspend-core-2.0.0.97/kernel/power/suspend2.c software-suspend-core-2.0.0.98/kernel/power/suspend2.c
--- software-suspend-core-2.0.0.97/kernel/power/suspend2.c	2004-07-09 23:13:15.000000000 +1000
+++ software-suspend-core-2.0.0.98/kernel/power/suspend2.c	2004-07-09 23:13:22.000000000 +1000
@@ -522,7 +522,15 @@
 #if LINUX_VERSION_CODE > KERNEL_VERSION(2,5,0)
 	device_flush_retain_state();
 	drivers_suspend();
+
+	if (pm_ops && pm_ops->prepare) {
+		if (!pm_ops->prepare(PM_SUSPEND_DISK)) {
+			if (pm_ops && pm_ops->enter)
+				pm_ops->enter(PM_SUSPEND_DISK);
+		}
+	}
 #endif
+
 	sys_reboot(LINUX_REBOOT_MAGIC1, LINUX_REBOOT_MAGIC2,
 			LINUX_REBOOT_CMD_POWER_OFF, NULL);
 
@@ -594,6 +602,11 @@
 
 	spin_unlock_irqrestore(&suspend_irq_lock, suspendirqflags);
 
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2,5,0)
+	if (pm_ops && pm_ops->finish)
+		pm_ops->finish(PM_SUSPEND_DISK);
+#endif
+
 #ifdef DEFAULT_SUSPEND_CONSOLE
 	post_resume_console_redraw();
 #endif
@@ -1299,7 +1312,7 @@
 static int get_suspend_debug_info(void);
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2,5,0)
-extern inline int notify_resume(void);
+extern int notify_resume(void);
 #if defined(CONFIG_MAGIC_SYSRQ) && defined(CONFIG_VT)
 extern int sysrq_pressed;
 #endif
diff -ruN software-suspend-core-2.0.0.97/kernel/power/swapwriter.c software-suspend-core-2.0.0.98/kernel/power/swapwriter.c
--- software-suspend-core-2.0.0.97/kernel/power/swapwriter.c	2004-07-09 23:13:15.000000000 +1000
+++ software-suspend-core-2.0.0.98/kernel/power/swapwriter.c	2004-07-09 23:13:22.000000000 +1000
@@ -985,9 +985,8 @@
 			if (suspend_early_boot_message(
 					"Failed to get access to the "
 					"resume header device.\nYou could be "
-					"booting with a 2.4 kernel when you "
-					"suspended a 2.6 kernel or vice "
-					"versa."))
+					"booting with a 2.6 kernel when you "
+					"suspended a 2.4 kernel."))
 				swapwriter_invalidate_image();
 
 			return -EINVAL;
@@ -1019,8 +1018,8 @@
 		if (set_blocksize(header_block_device, headerblocksize)) {
 			if (suspend_early_boot_message("Failed to get access to the "
 				"resume header device.\nYou could be booting "
-				"with a 2.4 kernel when you suspended a 2.6 "
-				"kernel or vice versa."))
+				"with a 2.6 kernel when you suspended a 2.4 "
+				"kernel."))
 				swapwriter_invalidate_image();
 
 			return -EINVAL;
